src/Microsoft.Health.Fhir.CosmosDb/Features/Storage/FhirCosmosClientInitializer.cs

```diff
@@ -0,0 +1,207 @@
+                }
+
+                _logger.LogInformation("Cosmos DB Database {DatabaseId} and collections successfully initialized", cosmosDataStoreConfiguration.DatabaseId);
+            }
+            catch (Exception ex)
+            {
+                _logger.LogCritical(ex, "Cosmos DB Database {DatabaseId} and collections initialization failed", cosmosDataStoreConfiguration.DatabaseId);
+                throw;
+            }
+        }
+
+        private class FhirCosmosSerializer : CosmosSerializer
+        {
+            private readonly JsonSerializer _serializer = CreateSerializer();
+            private static readonly RecyclableMemoryStreamManager _manager = new RecyclableMemoryStreamManager();
+
+            private static JsonSerializer CreateSerializer()
+            {
+                // Setting TypeNameHandling to any value other than 'None' will be flagged
+                // as causing potential security issues
+                var serializerSettings = new JsonSerializerSettings
+                {
+                    NullValueHandling = NullValueHandling.Ignore,
+                    DateFormatHandling = DateFormatHandling.IsoDateFormat,
+                    DateParseHandling = DateParseHandling.DateTimeOffset,
+                    DateTimeZoneHandling = DateTimeZoneHandling.RoundtripKind,
+                    TypeNameHandling = TypeNameHandling.None,
+                };
+
+                serializerSettings.Converters.Add(new StringEnumConverter());
+
+                // By default, the Json.NET serializer uses 'F' instead of 'f' for fractions.
+                // 'F' will omit the trailing digits if they are 0. You might end up getting something like '2018-02-07T20:04:49.97114+00:00'
+                // where the fraction is actually 971140. Because the ordering is done as string,
+                // if the values don't always have complete 7 digits, the comparison might not work properly.
+                serializerSettings.Converters.Add(new IsoDateTimeConverter { DateTimeFormat = "o" });
+
+                return JsonSerializer.Create(serializerSettings);
+            }
+
+            public override T FromStream<T>(Stream stream)
+            {
+                using var textReader = new StreamReader(stream);
+                using var reader = new JsonTextReader(textReader);
+                return _serializer.Deserialize<T>(reader);
+            }
+
+            public override Stream ToStream<T>(T input)
+            {
+                MemoryStream stream = _manager.GetStream();
+                using var writer = new NonClosingStreamWriter(stream);
+                using var jsonWriter = new JsonTextWriter(writer);
+                _serializer.Serialize(jsonWriter, input);
+                jsonWriter.Flush();
+                stream.Position = 0;
+                return stream;
+            }
+        }
+
+        private class NonClosingStreamWriter : StreamWriter
```

Comment:
You can instead specify that a StreamWriter leave the stream open by a constructor parameter.